### ISSUE TITLE
Containers: make opening a container from Docker file a nice experience

### ISSUE THREAD

**Issue by bpasero** [microsoft/vscode-remote-release#2468]
Occasionally I need to test something in VSCode's docker container for Linux that we use for building VSCode: https://github.com/microsoft/vscode-linux-build-agent/tree/master/x64

That is the moment when I am using the Remote Container and Docker extension. The flow I want is the following: given a `Dockerfile`, allow me to open a local folder on my OS inside that docker file. This should be a one-click command that does:
* create an image
* create a container
* start the container
* opens a folder I choose in that container in a VSCode instance

Today this is not possible, the steps are:
* right-click `Dockerfile` -> create image
* open a terminal and run the image so that it becomes a container to connect to (it needs to run a long running command to not shutdown directly)
* right click container -> open folder in container

**Comment by egamma** [comment]
@bpasero this looks to me like the 'inspecting containers' scenario that is described here https://code.visualstudio.com/blogs/2019/10/31/inspecting-containers (the blog is using docker_compose, but the same flow applies to a single docker image). Have you tried:

* right-click `Dockerfile` -> create image
* run the image 
* attach to the container from the Docker Viewlet's context menu 'Attach to Visual Studio Code' or from the Remote Explorer's Containers section.

**Comment by bpasero** [comment]
@egamma yeah, but hard to understand. The terms image, container etc. make sense for someone that knows Docker. For someone like me that does not understand these terms, it would be better to have a flow where VSCode would do everything for me. 

For example, what is Run vs Run Interactive? And when I do "Run Interactive" and close the first window, the second window is suddenly disconnected? 

Please allow: Right-click `Dockerfile` with an entry to open a folder within and do these magic things under the hood.

PS: It does not help that Docker extension sorts "Containers" to the top and then "Images". It should be the other way around because the hierarchy is Image -> Container.

**Comment by bpasero** [comment]
Maybe the "Docker" extension should stay as is and be for people that actually know Docker. Our extension however could optimise for everything that is "Open with VSCode" related. E.g. here, add an entry to open a dockerfile as container:

![image](https://user-images.githubusercontent.com/900690/75753867-f09cb480-5d2b-11ea-8c34-015a23778539.png)

**Comment by bpasero** [comment]
Actually I should not be required to install Docker extension to connect remotely...

**Comment by egamma** [comment]
>Actually I should not be required to install Docker extension to connect remotely...

yes, you can now do this all from the Remote Explorer>Containers section.

**Comment by bpasero** [comment]
> Since you are using a Dockerfile scenario, it is fair that the user has some understanding of Docker and is familiar with running a container from the CLI.

For many maybe, not for me. Just expressing my pain points. After all the one that created the `Dockerfile` (Joao) may not be the same person as the one triaging issues (me).

**Comment by egamma** [comment]
>Whenever I am using remote I am constantly hitting this wall of having to read into docs on other places outside the tool. That should not be the case ideally. We can do better.

Makes sense & agree we should look into improving this.

**Comment by chrmarti** [comment]
Note that there are many options along the way, when building the image, when creating the container. That's why we introduced the `devcontainer.json`. Having an action that goes from the `Dockerfile` to an up-and-running container with VS Code attached will only work for the simple cases.

On `Run` and `Run Interactive` removing the container: https://github.com/microsoft/vscode-docker/issues/1217
